rfcs/0010-define-calculations-of-existential-deposit.md

```diff
@@ -0,0 +1,163 @@
+```
+
+This method of setting ED ensures requirement 2.
+
+Example
+```
+asset_a_local_ed = 100_000_000
+asset_a_ratio = Ratio(1_000_000_000, 1_000_000_000_000)
+asset_b_local_ed = 100_000
+asset_a_ratio = Ratio(1_000_000, 1_000_000_000_000)
+native_asset_ed = 100_000_000_000
+
+fn native_to_asset(native_balance, asset_ratio) {
+  native_balance * asset_ratio
+}
+
+fn asset_to_native(asset_balance, asset_ratio) {
+  asset_balance / asset_ratio
+}
+
+asset_a_ed_as_native = asset_to_native(asset_a_local_ed, asset_a_ratio)
+asset_b_ed_as_native = asset_to_native(asset_b_local_ed, asset_b_ratio)
+native_as_asset_b = native_to_asset(native_asset_ed, asset_b_ratio)
+asset_a_as_asset_b = native_to_asset(asset_a_ed_as_native, asset_b_ratio)
+
+assert(asset_a_ed_as_native == native_asset_ed)
+assert(asset_b_ed_as_native == native_asset_ed)
+assert(asset_a_ed_as_native == asset_b_ed_as_native)
+assert(native_as_asset_b == asset_a_as_asset_b)
+```
+
+### Well-Known Assets & Asset Registry Storage
+
+For a asset to have its ED be correctly calculated, and therefore to exists on 
+chain, it should have its `foreign_asset_ratio` defined in either our Well-Known 
+assets list, or stored in Asset Registry.
+
+## Implementation
+
+### Updates to the Functionality of `multi_existential_deposits`
+
+The current implementation of `multi_existential_deposits` does the following:
+  
+  1. **If an ED is defined for an asset in Asset Registry, return the ED as 
+     is.**
+  
+  2. **Otherwise, pull a hard-coded ED from a `match` statement**.
+  
+  3. If no matches are found, return `Balance::MAX` which will prune unknown
+     assets.
+    
+It should be updated as follows:
+
+  1. **If a asset ratio is defined for an asset in Asset Registry or as a 
+     well-known asset, calculate the standard ED as defined by this document.**
+  
+  3. If no matches are found, return `Balance::MAX` which will prune unknown
+     assets.
+    
+### Existential Deposit Retrieval
+
+For use in the front-end, an RPC will need to be made available that exposes the
+functionality of `multi_existential_deposits`.
+    
+### Asset Registry
+
+**Note**: If AssetRegistry storage is already in use on Picasso, we will need to
+perform a storage migration in the next runtime upgrade for all of these changes
+to be possible.
+
+* The function `AssetRegistry::register_asset` should be updated with the 
+  following signature:
+
+  ```rust
+  pub fn register_asset(
+    origin: OriginFor<T>,
+    location: T::ForeignAssetId,
+    ratio: Ratio, // Used to be optional
+    decimals: Option<Exponent>,
+  ) -> DispatchResultWithPostInfo;
+  ```
+
+* Additionally, Asset Registry needs to support creating assets with a defined
+asset ID. 
+
+  ```rust
+  pub fn register_asset_with_id(
+    origin: OriginFor<T>,
+    asset_id: T::AssetId,
+    location: T::ForeignAssetId,
+    ratio: Ratio, // Used to be optional
+    decimals: Option<Exponent>,
+  ) -> DispatchResultWithPostInfo;
+  ```
```

Comment:
Looking back at this, I don't think this will be needed for Albula.

We no longer wish to directly overwrite the ED of well-known assets and with @andresmechali 's work, we don't need to worry about exposing every well-known asset through Asset Registry.

The only other use case I could see for this would be overwriting the ratios of well-known assets with Asset Registry.

Thoughts? @vimukthi-git @dzmitry-lahoda

Reply:
Yes the idea is to hard code assets anyway.